### PR TITLE
preparation for distinguishing bad input from bad api usage

### PR DESCRIPTION
**Pull request by jonsneyers**
Currently there is no distinction in the decoder API between errors caused by bad API usage (e.g. calling the functions in the wrong order, or with arguments that make no sense) and errors caused by bad input (corrupt/invalid/unsupported bitstreams).

Since that is probably a useful distinction to make (the former is an application bug while the latter is something that can happen at runtime and needs to be handled by the application), as a start I'm adding a macro JXL_INPUT_ERROR in addition to the JXL_API_ERROR macro that already existed, and changed all instances where the error is not an API usage error to use this macro instead. For now they both still do the same thing (return JXL_DEC_ERROR) but this will make it easier if we later want to have some more fine-grained error mechanism.

Flyby: the box size overflow check was not quite correct (the box header was counted twice), so fixing/simplifying that a bit.

### CODE REVIEW

File: lib/jxl/decode.cc [hunk 1]
```diff
@@ -1269,11 +1265,11 @@ JxlDecoderStatus JxlDecoderProcessCodestream(JxlDecoder* dec) {
         // No overflow, checked in CheckSizeLimit.
         size_t num_pixels = frame_dim.xsize * frame_dim.ysize;
         if (dec->used_cpu_base + num_pixels < dec->used_cpu_base) {
-          return JXL_API_ERROR("used too much CPU");
+          return JXL_INPUT_ERROR("used too much CPU");
```
Comment (szabadka) [inline lib/jxl/decode.cc:1268]: I think something like "image too large" would be better here.

File: lib/jxl/decode.cc [hunk 2]
```diff
@@ -2047,18 +2043,18 @@ JxlDecoderStatus JxlDecoderProcessInput(JxlDecoder* dec) {
   JxlDecoderStatus status = HandleBoxes(dec);
 
   if (status == JXL_DEC_NEED_MORE_INPUT && dec->input_closed) {
-    return JXL_API_ERROR("missing input");
+    return JXL_INPUT_ERROR("missing input");
```
Comment (szabadka) [inline lib/jxl/decode.cc:2046]: this should be more generic, something like "premature end of input"

File: lib/jxl/decode.cc [hunk 3]
```diff
@@ -2270,7 +2266,8 @@ JxlDecoderStatus JxlDecoderGetColorAsICCProfile(const JxlDecoder* dec,
   JxlDecoderStatus status =
       JxlDecoderGetICCProfileSize(dec, target, &wanted_size);
   if (status != JXL_DEC_SUCCESS) return status;
-  if (size < wanted_size) return JXL_API_ERROR("ICC profile output too small");
+  if (size < wanted_size)
+    return JXL_INPUT_ERROR("ICC profile output too small");
```
Comment (szabadka) [inline lib/jxl/decode.cc:2270]: This is actually an API error, we get here if the buffer provided in the API call was too small.

File: lib/jxl/decode.cc [hunk 4]
```diff
@@ -1269,11 +1265,11 @@ JxlDecoderStatus JxlDecoderProcessCodestream(JxlDecoder* dec) {
         // No overflow, checked in CheckSizeLimit.
         size_t num_pixels = frame_dim.xsize * frame_dim.ysize;
         if (dec->used_cpu_base + num_pixels < dec->used_cpu_base) {
-          return JXL_API_ERROR("used too much CPU");
+          return JXL_INPUT_ERROR("used too much CPU");
         }
         dec->used_cpu_base += num_pixels;
         if (dec->used_cpu_base > dec->cpu_limit_base) {
-          return JXL_API_ERROR("used too much CPU");
+          return JXL_INPUT_ERROR("used too much CPU");
```
Comment (szabadka) [inline lib/jxl/decode.cc:1272]: "image too large"

File: lib/jxl/decode.cc [hunk 5]
```diff
@@ -2303,7 +2300,7 @@ JxlDecoderStatus PrepareSizeCheck(const JxlDecoder* dec,
   *bits = BitsPerChannel(format->data_type);
 
   if (*bits == 0) {
-    return JXL_API_ERROR("Invalid/unsupported data type");
+    JXL_API_ERROR("Invalid/unsupported data type");
```
Comment (szabadka) [inline lib/jxl/decode.cc:2303]: why remove the return?

Comment (jonsneyers) [reply lib/jxl/decode.cc]: oops, putting it back

File: lib/jxl/decode.cc (hunk 3 above)
Comment (jonsneyers) [reply lib/jxl/decode.cc]: right, changing it

**Review by szabadka — APPROVED** [review]

**Pull request merged**